contrib/autopilot/c-lightning-autopilot.py

```diff
@@ -0,0 +1,269 @@
+generated by using the diverse strategy to mix the 4 heuristics. 
+
+Currently the software will not check, if sufficient funds are available
+or if a channel already exists.
+'''
+
+from os.path import expanduser
+import argparse
+import logging
+import math
+import pickle
+import sys
+
+from lightning import LightningRpc
+import dns.resolver
+
+from bech32 import bech32_decode, CHARSET, convertbits
+from lib_autopilot import Autopilot
+from lib_autopilot import Strategy
+import networkx as nx
+
+
+class CLightning_autopilot(Autopilot):
+     
+    def __init__(self, path, input=None, dont_store=None):
+        self.__add_clogger()
+        
+        self.__rpc_interface = LightningRpc(path)
+        self.__clogger.info("connection to RPC interface successful")
+        
+        G = None
+        if input:
+            try:
+                self.__clogger.info(
+                    "Try to load graph from file system at:" + input)
+                with open(input, "rb") as infile:
+                    G = pickle.load(infile)
+                    self.__clogger.info(
+                        "Successfully restored the lightning network graph from data/networkx_graph")
+            except FileNotFoundError:
+                self.__clogger.info(
+                    "input file not found. Load the graph from the peers of the lightning network")
+                G = self.__download_graph()
+        else:
+            self.__clogger.info(
+                    "no input specified download graph from peers")
+            G = self.__download_graph()
+        
+        if dont_store is None:          
+            with open("lightning_networkx_graph.pickle", "wb") as outfile:
+                    pickle.dump(G, outfile, pickle.HIGHEST_PROTOCOL)
+
+        Autopilot.__init__(self,G)
+
+            
+        
+    def __add_clogger(self):
+        """ initiates the logging service for this class """
+        # FIXME: adapt to the settings that are proper for you
+        self.__clogger = logging.getLogger('clightning-autopilot')
+        self.__clogger.setLevel(logging.INFO)
+        ch = logging.StreamHandler()
+        ch.setLevel(logging.INFO)
+        formatter = logging.Formatter(
+            '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        ch.setFormatter(formatter)
+        self.__clogger.addHandler(ch)
+        self.__clogger.info("set up logging infrastructure")
+    
+    def __get_seed_keys(self):
+        """
+        retrieve the nodeids of the ln seed nodes from lseed.bitcoinstats.com
+        """
+        domain = "lseed.bitcoinstats.com"
+        srv_records = dns.resolver.query(domain,"SRV")
+        res = []
+        for srv in srv_records:
+            bech32 = str(srv.target).rstrip(".").split(".")[0]
+            data = bech32_decode(bech32)[1]
+            decoded = convertbits(data, 5, 4)
+            res.append("".join(
+                ['{:1x}'.format(integer) for integer in decoded])[:-1])
```

Comment:
this code should decode the retrieved subdomains of lseeds.bitcoinstats.com to bech32 addresses so that the autopilot can peer with them. Is there an easier way to do it? the last line in particular transfers the hex values of the decoded address to a string since strings are needed for the rpc-interface connect call

Reply:
My comment is mainly that it does a lot of magic (dropping the last letter with `[:-1]`, using `{:1x}` to format, and similar). If you could split it into multiple operations with some additional comments as to what is happening it'd be easier to follow along.

The rule usually is that list comprehensions are ok, if they clear about what they do and they don't span more than one line, this breaks both of these rules :wink: